services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/service/session/Session.java

```diff
@@ -0,0 +1,183 @@
+ */
+
+package org.openkilda.floodlight.service.session;
+
+import org.openkilda.floodlight.error.SessionCloseException;
+import org.openkilda.floodlight.error.SessionConnectionLostException;
+import org.openkilda.floodlight.error.SessionErrorResponseException;
+import org.openkilda.floodlight.error.SessionRevertException;
+import org.openkilda.floodlight.error.SwitchWriteException;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import net.floodlightcontroller.core.IOFSwitch;
+import org.projectfloodlight.openflow.protocol.OFBarrierRequest;
+import org.projectfloodlight.openflow.protocol.OFErrorMsg;
+import org.projectfloodlight.openflow.protocol.OFMessage;
+import org.projectfloodlight.openflow.protocol.OFType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+public class Session implements AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(Session.class);
+
+    private final SwitchSessions group;
+    private final IOFSwitch sw;
+
+    private CompletableFuture<Optional<OFMessage>> closingBarrier;
+    private boolean error = false;
+    private boolean completed = false;
+
+    private final Map<Long, CompletableFuture<Optional<OFMessage>>> requestsByXid = new ConcurrentHashMap<>();
+
+    Session(SwitchSessions group, IOFSwitch sw) {
+        this.group = group;
+        this.sw = sw;
+    }
+
+    /**
+     * Send OF message to the switch and register it in session to trace possible responses.
+     */
+    public CompletableFuture<Optional<OFMessage>> write(OFMessage message) throws SwitchWriteException {
+        ensureOpen();
+
+        CompletableFuture<Optional<OFMessage>> future = prepareRequest(message);
+        actualWrite(future, message);
+
+        return future;
+    }
+
+    public void resetError() {
+        error = false;
+    }
+
+    @Override
+    public void close() throws SwitchWriteException {
+        if (error) {
+            SessionRevertException e = new SessionRevertException(sw.getId());
+            incompleteRequestsStream()
+                    .forEach(entry -> entry.completeExceptionally(e));
+            return;
+        }
+
+        if (closingBarrier != null) {
+            throw new IllegalStateException("Session already closed");
+        }
+
+        OFBarrierRequest barrier = sw.getOFFactory().barrierRequest();
+        closingBarrier = prepareRequest(barrier);
+        try {
+            actualWrite(closingBarrier, barrier);
+        } catch (SwitchWriteException e) {
+            SessionCloseException closeError = new SessionCloseException(sw.getId());
+            incompleteRequestsStream()
+                    .forEach(entry -> entry.completeExceptionally(closeError));
+            throw e;
+        }
```

Comment:
Shouldn't we validate here that the session is empty / all expected responses received?